### PR TITLE
Sync Summary Details

### LINKED ISSUE #42: Improve styles of TaskSummary
- [ ] Improve styles of TaskSummary.

Previous Tasks:

- [x] Create a modal that outputs the sync summary.

- [x] Output raw JSON response.

- [X] Handle 4XX errors.

- [x] Create more ExpandablePanels for each destination instace.

- [x] In each destination instance add a [TaskSummary](https://github.com/dhis2/import-export-app/tree/master/src/components/TaskSummary) of deleted/updated/created.

- [x] Properly output specific issues with metadata found.

- [x] Add a logger that stores the summaries in the dataStore.

- [x] Create a history page that reads outputs all the recent synchronizations.

### PR DESCRIPTION
**Pull request by SferaDev**
### :pushpin: References

* **Issue:** Closes #42 

### :memo: Implementation

- Added the main logic from the import-export-app for reporting.

- Improved the import-export-app UI into material-ui tables.

- Collapsed all when more than one instance involved.

- Solved "Warning contextType"

- Solved "Network error causing to finish sync"

- Solved "Dropdown label was not defined"

### :art: Screenshots

image

image

image

image

image

### :fire: Is there anything the reviewer should know to test it?

- The two tables generated might be combined later into a parametric single method.

- History page (logger) should be moved from #42 into its own card

### CODE REVIEW

File: src/components/sync-configurator/BaseSyncConfigurator.jsx [hunk 1]
```diff
@@ -173,4 +173,4 @@ class BaseSyncConfigurator extends React.Component {
     }
 }
 
-export default withRouter(withSnackbar(withStyles(styles)(BaseSyncConfigurator)));
+export default withSnackbar(withRouter(withStyles(styles)(BaseSyncConfigurator)));
```
Comment (tokland) [inline src/components/sync-configurator/BaseSyncConfigurator.jsx]: I guess that solves the "Warning contextType". Did you understand the details?

File: src/components/sync-summary/SyncSummary.jsx [hunk 2]
```diff
@@ -2,13 +2,38 @@ import React from "react";
...
 import PropTypes from "prop-types";
 import { ConfirmationDialog } from "d2-ui-components";
+import ReactJson from "react-json-view";
+
+import { withStyles } from "@material-ui/core";
+import Table from "@material-ui/core/Table";
+import TableHead from "@material-ui/core/TableHead";
+import TableRow from "@material-ui/core/TableRow";
+import TableCell from "@material-ui/core/TableCell";
+import TableBody from "@material-ui/core/TableBody";
```
Comment (tokland) [inline src/components/sync-summary/SyncSummary.jsx]: Minor/irrelevant: Vscode tends to prefer `import { Table, TableHeader, ... } from "@material-ui/core";`, which looks a bit more compact.

File: src/components/sync-summary/SyncSummary.jsx [hunk 3]
```diff
@@ -35,10 +114,60 @@ class SyncSummary extends React.Component {
...
+                                </ExpansionPanelSummary>
+
+                                <ExpansionPanelDetails className={classes.expansionPanelDetails}>
+                                    <Typography variant={"overline"}>Summary</Typography>
+                                </ExpansionPanelDetails>
+
+                                <ExpansionPanelDetails className={classes.expansionPanelDetails}>
+                                    {SyncSummary.buildSummaryTable([
+                                        ...responseElement.report.typeStats,
+                                        { type: "Total", ...responseElement.stats },
```
Comment (tokland) [inline src/components/sync-summary/SyncSummary.jsx]: Is it easy / makes sense to highlight the final total row with a background color?

File: src/components/sync-summary/SyncSummary.jsx [hunk 4]
```diff
@@ -35,10 +114,60 @@ class SyncSummary extends React.Component {
...
+                                            responseElement.instance.url
+                                        })`}
+                                    </Typography>
+                                    <Typography className={classes.expansionPanelHeading2}>
+                                        {`Status: ${responseElement.status}`}
+                                    </Typography>
+                                </ExpansionPanelSummary>
+
+                                <ExpansionPanelDetails className={classes.expansionPanelDetails}>
+                                    <Typography variant={"overline"}>Summary</Typography>
```
Comment (tokland) [inline src/components/sync-summary/SyncSummary.jsx]: Untranslated literal

File: src/components/sync-summary/SyncSummary.jsx [hunk 5]
```diff
@@ -35,10 +114,60 @@ class SyncSummary extends React.Component {
...
+                                key={`row-${i}`}
+                            >
+                                <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
+                                    <Typography className={classes.expansionPanelHeading1}>
+                                        {`${responseElement.instance.name} (${
+                                            responseElement.instance.url
+                                        })`}
+                                    </Typography>
+                                    <Typography className={classes.expansionPanelHeading2}>
+                                        {`Status: ${responseElement.status}`}
```
Comment (tokland) [inline src/components/sync-summary/SyncSummary.jsx]: Untranslated literal

File: src/components/sync-summary/SyncSummary.jsx [hunk 6]
```diff
@@ -35,10 +114,60 @@ class SyncSummary extends React.Component {
...
+                                        { type: "Total", ...responseElement.stats },
+                                    ])}
+                                </ExpansionPanelDetails>
+
+                                {responseElement.report.messages.length > 0 && (
+                                    <div>
+                                        <ExpansionPanelDetails
+                                            className={classes.expansionPanelDetails}
+                                        >
+                                            <Typography variant={"overline"}>Messages</Typography>
```
Comment (tokland) [inline src/components/sync-summary/SyncSummary.jsx]: Untranslated literal

Comment (tokland) [inline src/components/sync-summary/SyncSummary.jsx]: Can use direct `variant="overline"`.

File: src/components/sync-summary/SyncSummary.jsx [hunk 7]
```diff
@@ -35,10 +114,60 @@ class SyncSummary extends React.Component {
...
+                                    </div>
+                                )}
+                            </ExpansionPanel>
+                        ))}
+
+                        <ExpansionPanel>
                             <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
-                                <Typography>JSON Response</Typography>
+                                <Typography className={classes.expansionPanelHeading1}>
+                                    JSON Response
```
Comment (tokland) [inline src/components/sync-summary/SyncSummary.jsx]: Untranslated

File: src/utils/d2.ts [hunk 8]
```diff
@@ -58,3 +58,7 @@ export function cleanModelName(d2: D2, id: string, caller: string): string | nul
         return null;
     }
 }
+
+export function getClassName(className: string): string {
+    return className.substr(className.lastIndexOf(".") + 1);
```
Comment (tokland) [inline src/utils/d2.ts]: Proposal: `_(className).split(".").last()`

File: src/components/sync-configurator/BaseSyncConfigurator.jsx (hunk 1 above)
Comment (SferaDev) [reply src/components/sync-configurator/BaseSyncConfigurator.jsx]: More or less I understand the cause but not the details. It's something about the way withRouter builds the HOC, it doesn't like having contextType set inside it.

File: src/components/sync-summary/SyncSummary.jsx (hunk 3 above)
Comment (SferaDev) [reply src/components/sync-summary/SyncSummary.jsx]: This is for @adrianq :)

I guess making it bold will suffice.

Comment (adrianq) [reply src/components/sync-summary/SyncSummary.jsx]: Although we all love diaphanous interface some styling would be highly appreciated here I agree ;) Anyway, let's keep it like for now. We can rethink it a bit as part of https://github.com/EyeSeeTea/metadata-synchronization/issues/42

**Comment by adrianq** [comment]
Thanks for this guys!

**Pull request merged**